File: python/2021/day3/day3.py
#!/usr/bin/env python3
#import solution

with open('./day3.in', 'r') as file:
    data = file.read().splitlines()

def get_common(bit, reversed = False):
    more_zeroes = bit.count('0') > bit.count('1')
    if reversed:
        return '1' if more_zeroes else '0'
    return '0' if more_zeroes else '1'

def bitwise_not(bits):
    return ''.join(map(lambda x: '1' if x == '0' else '0', bits))

def get_rating(data, reversed = False, bit_pos = 0):
    valid = list([x for x in data if x[bit_pos] == get_common(list(zip(*data))[bit_pos], reversed)])
    if len(valid) == 1: 
        return valid[0]
    return get_rating(valid, reversed, bit_pos + 1)


common_bits = ''.join(map(get_common, zip(*data)))
print(int(common_bits, 2) * int(bitwise_not(common_bits), 2))
print(int(get_rating(data), 2) * int(get_rating(data, True), 2))

File: cpp/2021/day3/day3.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std; 

inline int bin(string& str) {
    return stoi(str, 0, 2);
}

template<typename Lambda>
int get_rating(vector<string> data, Lambda compare, int bit_pos = 0) {
    if (data.size() == 1) return bin(data[0]);

    vector<string> valid;
    //find which bit is most common
    int counter = 0; // == 0 means same, < 0 means more '0', > 0 means more '1'
    for (int row = 0; row < data.size(); row++) {
        counter += (data[row][bit_pos] == '1') ? 1 : -1;
    }

    for (string &s : data) {
        if ('0' + compare(counter) == s[bit_pos]) 
            valid.push_back(s);
    }

    return get_rating(valid, compare, bit_pos+1);
}

int main() {

    ifstream infile;
    infile.open("./day3.in");
    vector<string> data;
    string line;

    while (getline(infile, line)) {
        data.push_back(line);
    }

    infile.close();

    string gamma = "", epsilon = "";

    for (int col = 0; col < line.length(); col++) {
       int counter = 0; // == 0 means same, < 0 means more '0', > 0 means more '1'
        for (int row = 0; row < data.size(); row++) {
            counter += (data[row][col] == '1') ? 1 : -1;
        }
        gamma += '0' + (counter >= 0);
        epsilon += '0' + (counter < 0);
    }

    cout << (bin(gamma) * bin(epsilon)) << endl;
    cout << get_rating(data, [](int x){ return (x >= 0); }) * get_rating(data, [](int x){ return (x < 0); }) << endl;

    return 0;
}